README
NAME
    String::Format - sprintf-like string formatting capabilities with
    arbitrary format definitions

ABSTRACT
    String::Format allows for sprintf-style formatting capabilities with
    arbitrary format definitions

SYNOPSIS
      use String::Format;

      my %fruit = (
            'a' => "apples",
            'b' => "bannanas",
            'g' => "grapefruits",
            'm' => "melons",
            'w' => "watermelons",
      );

      my $format = "I like %a, %b, and %g, but not %m or %w.";

      print stringf($format, %fruit);
  
      # prints:
      # I like apples, bannanas, and grapefruits, but not melons or watermelons.

DESCRIPTION
    String::Format lets you define arbitrary printf-like format sequences to
    be expanded. This module would be most useful in configuration files and
    reporting tools, where the results of a query need to be formatted in a
    particular way. It was inspired by mutt's index_format and related
    directives (see
    <URL:http://www.mutt.org/doc/manual/manual-6.html#index_format>).

FUNCTIONS
  stringf
    String::Format exports a single function called stringf. stringf takes
    two arguments: a format string (see FORMAT STRINGS, below) and a
    reference to a hash of name => value pairs. These name => value pairs
    are what will be expanded in the format string.

FORMAT STRINGS
    Format strings must match the following regular expression:

      qr/
         (%             # leading '%'
          (-)?          # left-align, rather than right
          (\d*)?        # (optional) minimum field width
          (?:\.(\d*))?  # (optional) maximum field width
          ({.*?})?      # (optional) stuff inside
          (\S)          # actual format character
         )/x;

    If the escape character specified does not exist in %args, then the
    original string is used. The alignment, minimum width, and maximum width
    options function identically to how they are defined in sprintf(3) (any
    variation is a bug, and should be reported).

    Note that Perl's sprintf definition is a little more liberal than the
    above regex; the deviations were intentional, and all deal with numeric
    formatting (the #, 0, and + leaders were specifically left out).

    The value attached to the key can be a scalar value or a subroutine
    reference; if it is a subroutine reference, then anything between the
    '{' and '}' ($5 in the above regex) will be passed as $_[0] to the
    subroutine reference. This allows for entries such as this:

      %args = (
          d => sub { POSIX::strftime($_[0], localtime) }, 
      );

    Which can be invoked with this format string:

      "It is %{%M:%S}d right now, on %{%A, %B %e}d."

    And result in (for example):

      It is 17:45 right now, on Monday, February 4.

    Note that since the string is passed unmolested to the subroutine
    reference, and strftime would Do The Right Thing with this data, the
    above format string could be written as:

      "It is %{%M:%S right now, on %A, %B %e}d."

    By default, the formats 'n', 't', and '%' are defined to be a newline,
    tab, and '%', respectively, if they are not already defined in the
    hashref of arguments that gets passed it. So we can add carriage returns
    simply:

      "It is %{%M:%S right now, on %A, %B %e}d.%n"

    Because of how the string is parsed, the normal "\n" and "\t" are turned
    into two characters each, and are not treated as a newline and tab. This
    is a bug.

FACTORY METHOD
    String::Format also supports a class method, named stringfactory, which
    will return reference to a "primed" subroutine. stringfatory should be
    passed a reference to a hash of value; the returned subroutine will use
    these values as the %args hash.

      my $self = Some::Groovy::Package->new($$, $<, $^T);
      my %formats = (
            'i' => sub { $self->id      },
            'd' => sub { $self->date    },
            's' => sub { $self->subject },
            'b' => sub { $self->body    },
      );
      my $index_format = String::Format->stringfactory(\%formats);

      print $index_format->($format1);
      print $index_format->($format2);

    This subroutine reference can be assigned to a local symbol table entry,
    and called normally, of course:

      *reformat = String::Format->stringfactory(\%formats);

      my $reformed = reformat($format_string);

AUTHOR
    darren chamberlain <darren@cpan.org>



Changes
Revision history for {{$dist->name}}

{{$NEXT}}

1.235     2022-12-31 16:49:41-05:00 America/New_York
        - update metadata

1.234     2021-08-28 17:28:18-04:00 America/New_York
        - move bench.pl from dist root so it won't get installed to @INC
        - fix some problems with doc layout

0.102084  2013-11-08 22:30:46 America/New_York
        - typo fixes
        - update repo and bugtracker links

0.102082  2010-10-19 14:34:34 America/New_York
          add missing prereq for Test::More

0.102081  2010-10-17 13:40:26 America/New_York
          allow subclasses to provide default_codes

          implement %% in a simpler, but not quite perl-like, way

0.102080  2010-07-27 22:07:11 America/New_York
          input hunkers have been tweaked and more fully documented

0.101620  2010-06-11 21:31:26 America/New_York
          add indexed_replace, require_arrayref_input, and indexed_stringf

0.100720  2010-03-13 09:36:03 America/New_York
          fix an ancillary file that brought in a bunch of bogus prereqs

0.100690  2010-03-10 21:09:36 America/New_York
          fix a 5.10-ism that slipped in

0.100680  2010-03-09 20:54:11 America/New_York
          bug fix: newlines no longer end formatting, but not allowed in braces
          (thanks for bug report, Christopher J. Madsen)

0.093221  2009-11-18 08:13:22 America/New_York
          remove stupidly left-in hardcoded version!

0.093220  2009-11-18 08:12:10 America/New_York
          fix a problem with prereqs and tests

0.093210  2009-11-17
          add the cookbook
          add method importer

0.093200  2009-11-16
          first release!


bench.pl
#!perl -l
use strict;
use lib 'lib';

use Benchmark;
use String::Format;
use String::Formatter;
use Template;

my $hash = {
  a => 'apples',
  b => 'bananas',
};

my $fmt = String::Formatter->new({
  codes => $hash,
});

my $index_format = String::Format->stringfactory($hash);

my $tt2 = Template->new;

print $index_format->("I like to eat %a and %b.");
print $fmt->format("I like to eat %a and %b.");

$tt2->process(\'I like to eat [%a%] and [%b%].', $hash, \my $str);
print $str;

timethese(100_000, {
  dlc  => sub { $index_format->("I like to eat %a and %b.") },
  rjbs => sub { $fmt->format("I like to eat %a and %b.") },
  # tt2  => sub {
  #   $tt2->process(\'I like to eat [%a%] and [%b%].', $hash, \my $str);
  # },
  perl => sub { sprintf("I like to eat %s and %s.", qw(apples bananas)) },
});


lib/String/Stringf.pm
use strict;
use warnings;
package String::Stringf;
# ABSTRACT: build sprintf-like functions of your own

our $VERSION = '1.16';

=head1 SYNOPSIS

  use String::Stringf;

  my %fruit = (
    a => "apples",
    b => "bannanas",
    g => "grapefruits",
    m => "melons",
    w => "watermelons",
  );

  my $format = "I like %a, %b, and %g, but not %m or %w.";

  print stringf($format, %fruit);
  
...prints:

  I like apples, bannanas, and grapefruits, but not melons or watermelons.

=head1 DESCRIPTION

String::Stringf lets you define arbitrary printf-like format sequences to be
expanded.  This module would be most useful in configuration files and
reporting tools, where the results of a query need to be formatted in a
particular way.  String::Stringf is derived from String::Format, which was
inspired by mutt's index_format and related directives (see
<URL:http://www.mutt.org/doc/manual/manual-6.html#index_format>).

=cut

require 5.006;

use Params::Util ();
use Sub::Exporter -setup => {
  exports => [ stringf => \'_build_stringf' ],
  groups  => [ default => [qw(stringf)] ],
};

sub _replace {
  my (
    $letter, $orig,     $alignment, $min_width, $max_width,
    $passme, $formchar, $args,      $i_ref
  ) = @_;

  # For unknown escapes, return the orignial
  unless (defined $letter->{$formchar}) {
    Carp::croak("Unknown conversion in stringf-generated routine: %$formchar")
  }

  $alignment = '+' unless defined $alignment;

  my $replacement = $letter->{$formchar};
  if (ref $replacement eq 'CODE') {

    # $passme gets passed to subrefs.
    $passme ||= "";
    $passme =~ s/\A{//g;
    $passme =~ s/}\z//g;
    $replacement = $replacement->($passme, $$i_ref, $args);
  }

  my $replength = length $replacement;
  $min_width ||= $replength;
  $max_width ||= $replength;

  # length of replacement is between min and max
  if (($replength > $min_width) && ($replength < $max_width)) {
    return $replacement;
  }

  # length of replacement is longer than max; truncate
  if ($replength > $max_width) {
    return substr($replacement, 0, $max_width);
  }

  # length of replacement is less than min: pad
  if ($alignment eq '-') {

    # left align; pad in front
    return $replacement . " " x ($min_width - $replength);
  }

  # right align, pad at end
  return " " x ($min_width - $replength) . $replacement;
}

my $regex = qr/
 (%             # leading '%'
  (-)?          # left-align, rather than right
  (\d*)?        # (optional) minimum field width
  (?:\.(\d*))?  # (optional) maximum field width
  ({.*?})?      # (optional) stuff inside
  (\S)          # actual format character
 )
/x;

sub stringf {
  my $format = shift || return;
  my $args = Params::Util::_HASHLIKE($_[0]) ? shift : {@_};

  _build_stringf(__PACKAGE__, 'stringf', { formats => $args, },)->($format);
}

sub stringfactory {
  my $class = shift;
  my $args = Params::Util::_HASHLIKE($_[0]) ? shift : {@_};
  return $class->_build_stringf(stringf => { formats => $args });
}

sub _build_stringf {
  my ($self, $name, $arg) = @_;
  return $self->can('stringf') unless %$arg;
  Carp::confess('no formats given') unless my $format = $arg->{formats};

  $format->{'%'} = "%"  unless exists $format->{'%'};

  return sub {
    # This is the previous behavior, but I think we should die instead,
    # like sprintf. -- rjbs, 2009-05-15
    Carp::croak("not enough arguments for stringf-based format") unless @_;

    return unless defined (my $string = shift);
    my $i = -1;

    $string =~ s/$regex/
      $i++;
      _replace($format, $1, $2, $3, $4, $5, $6, \@_, \$i);
    /ge;

    return $string;
  };
}

1;
__END__

=head1 FUNCTIONS

=head2 stringf

String::Stringf can export a single function called stringf.  stringf
takes two arguments:  a format string (see FORMAT STRINGS, below) and
a reference to a hash of name => value pairs.  These name => value
pairs are what will be expanded in the format string.

=head1 FORMAT STRINGS

Format strings are strings with embedded format markers.  If you've used C or
Perl's C<sprintf>, you already understand.  If you haven't, you should read
about how C<sprintf> works.

Format markers are generally in the form:

  %    - a percent sign to begin the format
  ...  - (optional) various modifiers to the format like "-5" or "#" or "2$"
  {..} - (optional) a string inside braces
  s    - a short string (usually one character) identifying the conversion

Not all formatters found in Perl's C<sprintf> are yet supported.  Currently the
only format modifers must match:

    (-)?          # left-align, rather than right
    (\d*)?        # (optional) minimum field width
    (?:\.(\d*))?  # (optional) maximum field width

If a format string has an unknown conversion, an exception will be raised.

=head1 CONVERSIONS

C<stringf> routines must be provided with a set of conversions.  They're
provided as a hashref in which the keys are the conversion identifiers and the
values are either strings or coderefs.

If the values are strings then the string is the value that is used for the
conversion.  If it is a subroutine reference, the sub will be called once for
each time it appears in the format string, with the following arguments:

  $code->( ... );

This allows for entries such as this:

  %args = (
      d => sub { POSIX::strftime($_[0], localtime) }, 
  );

Which can be invoked with this format string:

  "It is %{%M:%S}d right now, on %{%A, %B %e}d."

And result in (for example):

  It is 17:45 right now, on Monday, February 4.

Note that since the string is passed unmolested to the subroutine
reference, and strftime would Do The Right Thing with this data, the
above format string could be written as:

  "It is %{%M:%S right now, on %A, %B %e}d."

By default, the formats 'n', 't', and '%' are defined to be a newline,
tab, and '%', respectively, if they are not already defined in the
hashref of arguments that gets passed it.  So we can add carriage
returns simply:

  "It is %{%M:%S right now, on %A, %B %e}d.%n"

Because of how the string is parsed, the normal "\n" and "\t" are
turned into two characters each, and are not treated as a newline and
tab.  This is a bug.

=begin :postlude

=head1 DERIVATION

String::Stringf is based on String::Format, written by Darren Chamberlain.  For
a history of the code, check the project's source code repository.  All bugs
should be reported to Ricardo Signes and String::Stringf.

=end :postlude


lib/String/Formatter/Cookbook.pm
use strict;
use warnings;
package String::Formatter::Cookbook;
# ABSTRACT: ways to put String::Formatter to use
1;

=encoding utf-8

=head1 OVERVIEW

String::Formatter is a pretty simple system for building formatting routines,
but it can be hard to get started without an idea of the sort of things that
are possible.

=head1 BASIC RECIPES

=head2 constants only

The simplest stringf interface you can provide is one that just formats
constant strings, allowing the user to put them inside other fixed strings with
alignment:

  use String::Formatter stringf => {
    input_processor => 'forbid_input',
    codes => {
      a => 'apples',
      b => 'bananas',
      w => 'watermelon',
    },
  };

  print stringf('I eat %a and %b but never %w.');

  # Output:
  # I eat apples and bananas but never watermelon.

If the user tries to parameterize the string by passing arguments after the
format string, an exception will be raised.

=head2 sprintf-like conversions

Another common pattern is to create a routine that behaves like Perl's
C<sprintf>, but with a different set of conversion routines.  (It will also
almost certainly have much simpler semantics than Perl's wildly complex
behavior.)

  use String::Formatter stringf => {
    codes => {
      s => sub { $_ },     # string itself
      l => sub { length }, # length of input string
      e => sub { /[^\x00-\x7F]/ ? '8bit' : '7bit' }, # ascii-safeness
    },
  };

  print stringf(
    "My name is %s.  I am about %l feet tall.  I use an %e alphabet.\n",
    'Ricardo',
    'ffffff',
    'abcchdefghijklllmnñopqrrrstuvwxyz',
  );

  # Output:
  # My name is Ricardo.  I am about 6 feet tall.  I use an 8bit alphabet.

B<Warning>: The behavior of positional string replacement when the conversion
codes mix constant strings and code references is currently poorly nailed-down.
Do not rely on it yet.

=head2 named conversions

This recipe acts a bit like Python's format operator when given a dictionary.
Rather than matching format code position with input ordering, inputs can be
chosen by name.

  use String::Formatter stringf => {
    input_processor => 'require_named_input',
    string_replacer => 'named_replace',

    codes => {
      s => sub { $_ },     # string itself
      l => sub { length }, # length of input string
      e => sub { /[^\x00-\x7F]/ ? '8bit' : '7bit' }, # ascii-safeness
    },
  };

  print stringf(
    "My %{which}s name is %{name}s.  My name is %{name}l letters long.",
    {
      which => 'first',
      name  => 'Marvin',
    },
  );

  # Output:
  # My first name is Marvin.  My name is 6 letters long.

Because this is a useful recipe, there is a shorthand for it:

  use String::Formatter named_stringf => {
    codes => {
      s => sub { $_ },     # string itself
      l => sub { length }, # length of input string
      e => sub { /[^\x00-\x7F]/ ? '8bit' : '7bit' }, # ascii-safeness
    },
  };

=head2 method calls

Some objects provide methods to stringify them flexibly.  For example, many
objects that represent timestamps allow you to call C<strftime> or something
similar.  The C<method_replace> string replacer comes in handy here:

  use String::Formatter stringf => {
    input_processor => 'require_single_input',
    string_replacer => 'method_replace',

    codes => {
      f => 'strftime',
      c => 'format_cldr',
      s => sub { "$_[0]" },
    },
  };

  print stringf(
    "%{%Y-%m-%d}f is also %{yyyy-MM-dd}c.  Default string is %s.",
    DateTime->now,
  );

  # Output:
  # 2009-11-17 is also 2009-11-17.  Default string is 2009-11-17T15:35:11.

This recipe is available as the export C<method_stringf>:

  use String::Formatter method_stringf => {
    codes => {
      f => 'strftime',
      c => 'format_cldr',
      s => sub { "$_[0]" },
    },
  };

You can easily use this to implement an actual stringf-like method:

  package MyClass;

  use String::Formatter method_stringf => {
    -as => '_stringf',
    codes => {
      f => 'strftime',
      c => 'format_cldr',
      s => sub { "$_[0]" },
    },
  };

  sub format {
    my ($self, $format) = @_;
    return _stringf($format, $self);
  }

=cut


Format.pm
package String::Stringf;

# ----------------------------------------------------------------------
#  Copyright (C) 2002,2009 darren chamberlain <darren@cpan.org>
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License as
#  published by the Free Software Foundation; version 2.
#
#  This program is distributed in the hope that it will be useful, but
#  WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
#  USA
# -------------------------------------------------------------------
require 5.006;
use strict;
use warnings;

use Params::Util (); # we actually get this for free with Sub::Exporter
use Sub::Exporter -setup => {
  exports => [ stringf => \'_build_stringf' ],
  groups  => [ default => [ qw(stringf) ] ],
};

our $VERSION = '1.16';

sub _replace {
    my ($letter, $orig, $alignment, $min_width,
        $max_width, $passme, $formchar, $args, $i_ref) = @_;

    # For unknown escapes, return the orignial
    unless (defined $letter->{$formchar}) {
      $$i_ref--;
      return $orig;
    }

    $alignment = '+' unless defined $alignment;

    my $replacement = $letter->{$formchar};
    if (ref $replacement eq 'CODE') {
        # $passme gets passed to subrefs.
        $passme ||= "";
        $passme =~ s/\A{//g;
        $passme =~ s/}\z//g;
        $replacement = $replacement->($passme, $$i_ref, $args);
    }

    my $replength = length $replacement;
    $min_width  ||= $replength;
    $max_width  ||= $replength;

    # length of replacement is between min and max
    if (($replength > $min_width) && ($replength < $max_width)) {
        return $replacement;
    }

    # length of replacement is longer than max; truncate
    if ($replength > $max_width) {
        return substr($replacement, 0, $max_width);
    }
    
    # length of replacement is less than min: pad
    if ($alignment eq '-') {
        # left align; pad in front
        return $replacement . " " x ($min_width - $replength);
    }

    # right align, pad at end
    return " " x ($min_width - $replength) . $replacement;
}

my $regex = qr/
               (%             # leading '%'
                (-)?          # left-align, rather than right
                (\d*)?        # (optional) minimum field width
                (?:\.(\d*))?  # (optional) maximum field width
                ({.*?})?      # (optional) stuff inside
                (\S)          # actual format character
             )/x;
sub stringf {
    my $format = shift || return;
    my $args = Params::Util::_HASHLIKE($_[0]) ? shift : { @_ };

    _build_stringf(
        __PACKAGE__,
        'stringf',
        {
          formats => $args,
        },
    )->($format);
}

sub stringfactory {
    my $class = shift;
    my $args = Params::Util::_HASHLIKE($_[0]) ? shift : { @_ };
    return $class->_build_stringf(stringf => { formats => $args });
}

sub _build_stringf {
    my ($self, $name, $arg) = @_;
    return $self->can('stringf') unless %$arg;
    Carp::confess('no formats given') unless my $format = $arg->{formats};

   $format->{'n'} = "\n" unless exists $format->{'n'};
   $format->{'t'} = "\t" unless exists $format->{'t'};
   $format->{'%'} = "%"  unless exists $format->{'%'};

    return sub {
        # This is the previous behavior, but I think we should die instead,
        # like sprintf. -- rjbs, 2009-05-15
        return unless defined $_[0];

        my $string = shift;
        my $i = -1;
        $string =~ s/$regex/
          $i++;
          _replace($format, $1, $2, $3, $4, $5, $6, \@_, \$i);
        /ge;
        return $string
    }
}

1;
__END__

=head1 NAME

String::Format - sprintf-like string formatting capabilities with
arbitrary format definitions

=head1 ABSTRACT

String::Format allows for sprintf-style formatting capabilities with
arbitrary format definitions

=head1 SYNOPSIS

  use String::Format;

  my %fruit = (
        'a' => "apples",
        'b' => "bannanas",
        'g' => "grapefruits",
        'm' => "melons",
        'w' => "watermelons",
  );

  my $format = "I like %a, %b, and %g, but not %m or %w.";

  print stringf($format, %fruit);
  
  # prints:
  # I like apples, bannanas, and grapefruits, but not melons or watermelons.

=head1 DESCRIPTION

String::Format lets you define arbitrary printf-like format sequences
to be expanded.  This module would be most useful in configuration
files and reporting tools, where the results of a query need to be
formatted in a particular way.  It was inspired by mutt's index_format
and related directives (see <URL:http://www.mutt.org/doc/manual/manual-6.html#index_format>).

=head1 FUNCTIONS

=head2 stringf

String::Format exports a single function called stringf.  stringf
takes two arguments:  a format string (see FORMAT STRINGS, below) and
a reference to a hash of name => value pairs.  These name => value
pairs are what will be expanded in the format string.

=head1 FORMAT STRINGS

Format strings must match the following regular expression:

  qr/
     (%             # leading '%'
      (-)?          # left-align, rather than right
      (\d*)?        # (optional) minimum field width
      (?:\.(\d*))?  # (optional) maximum field width
      ({.*?})?      # (optional) stuff inside
      (\S)          # actual format character
     )/x;

If the escape character specified does not exist in %args, then the
original string is used.  The alignment, minimum width, and maximum
width options function identically to how they are defined in
sprintf(3) (any variation is a bug, and should be reported).

Note that Perl's sprintf definition is a little more liberal than the
above regex; the deviations were intentional, and all deal with
numeric formatting (the #, 0, and + leaders were specifically left
out).

The value attached to the key can be a scalar value or a subroutine
reference; if it is a subroutine reference, then anything between the
'{' and '}' ($5 in the above regex) will be passed as $_[0] to the
subroutine reference.  This allows for entries such as this:

  %args = (
      d => sub { POSIX::strftime($_[0], localtime) }, 
  );

Which can be invoked with this format string:

  "It is %{%M:%S}d right now, on %{%A, %B %e}d."

And result in (for example):

  It is 17:45 right now, on Monday, February 4.

Note that since the string is passed unmolested to the subroutine
reference, and strftime would Do The Right Thing with this data, the
above format string could be written as:

  "It is %{%M:%S right now, on %A, %B %e}d."

By default, the formats 'n', 't', and '%' are defined to be a newline,
tab, and '%', respectively, if they are not already defined in the
hashref of arguments that gets passed it.  So we can add carriage
returns simply:

  "It is %{%M:%S right now, on %A, %B %e}d.%n"

Because of how the string is parsed, the normal "\n" and "\t" are
turned into two characters each, and are not treated as a newline and
tab.  This is a bug.

=head1 FACTORY METHOD

String::Format also supports a class method, named B<stringfactory>,
which will return reference to a "primed" subroutine.  stringfatory
should be passed a reference to a hash of value; the returned
subroutine will use these values as the %args hash.

  my $self = Some::Groovy::Package->new($$, $<, $^T);
  my %formats = (
        'i' => sub { $self->id      },
        'd' => sub { $self->date    },
        's' => sub { $self->subject },
        'b' => sub { $self->body    },
  );
  my $index_format = String::Format->stringfactory(\%formats);

  print $index_format->($format1);
  print $index_format->($format2);

This subroutine reference can be assigned to a local symbol table
entry, and called normally, of course:

  *reformat = String::Format->stringfactory(\%formats);

  my $reformed = reformat($format_string);

=head1 LICENSE

C<String::Format> is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; version 2.


=head1 AUTHOR

darren chamberlain <darren@cpan.org>
